### ISSUE TITLE
Announcements not signalled for method recompilations in subclasses when a superclass is changed

### ISSUE THREAD

**Issue by chisandrei** [guillep/pharo-core#15024]
**Bug description**

In some cases a change in the definition of a superclass leads to methods in subclasses being recompiled (which could be avoided) but no announcement is raised to indicate that subclass methods were recompiled. 

Consider the code below that creates the following hierarchy:
```
MyRootTestClass
  MySubClass1
    MySubClass12
      MySubClass123
MyOtherRootTestClass
```

```
Object subclass: #MyRootTestClass
	instanceVariableNames: '' 
	classVariableNames: ''
	package: 'MyPackage'.
	
Object subclass: #MyOtherRootTestClass
	instanceVariableNames: '' 
	classVariableNames: ''
	package: 'MyPackage'.

#MyRootTestClass asClass subclass: #MySubClass1
	instanceVariableNames: '' 
	classVariableNames: ''
	package: 'MyPackage'.

#MySubClass1 asClass subclass: #MySubClass12
	instanceVariableNames: '' 
	classVariableNames: ''
	package: 'MyPackage'.
#MySubClass12 asClass compile: 'two'.

targetMethodTwo := #MySubClass12 asClass >> #two.
self assert: targetMethodTwo isInstalled.

#MySubClass12 asClass subclass: #MySubClass123
	instanceVariableNames: '' 
	classVariableNames: ''
	package: 'MyPackage'.
#MySubClass123 asClass compile: 'three'.

targetMethodThree := #MySubClass123 asClass >> #three.
self assert: targetMethodThree isInstalled.
```

<img width="1088" alt="Screenshot 2023-10-16 at 16 16 01" src="https://github.com/pharo-project/pharo/assets/3534711/8b8166fb-757f-4613-be05-896a2f7fbbc3">

`MySubClass12` and `MySubClass123` have a method each.

Then we change the superclass of `MySubClass1` to `MyOtherRootTestClass`. This operations recompiled the two methods, `MySubClass12>>#two` and `MySubClass123>>#three`.

To double check we can also assert the identity hash of these methods:
```
self assert: targetMethodTwo isInstalled not.
self assert: (#MySubClass12 asClass >> #two) identityHash ~= targetMethodTwo identityHash.
```

```
self assert: targetMethodThree isInstalled not.
self assert: (#MySubClass123 asClass >> #three) identityHash ~= targetMethodThree identityHash.
```

However, there is no class changed announcement from the class builder that this compilation happened. We get only announcements related to the explicit change in `MySubClass1`, even if methods get recompiled in subclasses.

<img width="1088" alt="Screenshot 2023-10-16 at 16 19 38" src="https://github.com/pharo-project/pharo/assets/3534711/d6da4000-d646-4976-abd6-efa65d940ed4">

**Expected behavior**
We are not sure why those methods are recompiled, as there is no change to slots that would require recompilation of methods in subclasses. But if the recompilation happens, there should be at least a `ClassModificationApplied` announcement to signal that.

**Version information:**
Tested in Pharo 10, 11 and 12

**Expected development cost**
We think we could get the changed signalled correctly, if we could get the superclass chain of the old class and the new class in the `ShSuperclassChangedDetector` but we didn’t see how to get that. There could also be a workaround were we announce a class change every time we propagate a change to a subclass, but that does not sound like a proper fix.

**Comment by jecisc** [comment]
Hi,

I think I have the answer to some of the questions, but know that I am not 100% sure of what I'll say ;)

Those methods are recompiled because in case they are calling super, a literal needs to be updated in the method and thus we recompile the methods.

I think Marcus wants to improve this by recompiling only the methods using super (but it requires time :) ). 

About the announcement, we do not announce the recompilation during the update of a class because it would raise a lot of those announcements and the system would probably slow down too much. At some point we detected an announcement made in some cases during recompilation, and fixing this reduced the time of the Traits tests by 1min. And this was an edge case

**Comment by chisandrei** [comment]
In this case these methods have a completely empty implementation:

<img width="1088" alt="Screenshot 2023-10-16 at 17 01 49" src="https://github.com/pharo-project/pharo/assets/3534711/1a1682d0-9a79-40df-b74e-9b200e4f1dd3">

In this case not announcing these changes (while speeding up some operations), can lead to parts of the system relying on compiled methods reaching an inconsistent state. In our case, an index for speeding up senders and implementors in large images that stores compiled methods gets in an invalid state in case a method is recompiled in a class, and that recompilation is not announced...  

Would be quite useful for the system to not do these kinds of recompilations silently, without any chance for tools to be notified of such changes.

**Comment by MarcusDenker** [comment]
For super we do not need to recompile (the bytecode is always a "send and start lookup in the superclass of this class", both  for the old kind that takes the class from the last literal and the directedSuperSend: bc that takes it from the stack)

We need to only recompile if the superclass (recursively) has a different number of  instance variables, as then all the offsets have to be changed.

I think it is just an optimization not yet done: we should detect that the number of instance variables is the same and then not recompile.

**Comment by jecisc** [comment]
This is the number but also the order I guess? (Because we could change the index of the slots)

**Comment by MarcusDenker** [comment]
We do not announce recompile if the change is not relevant for the tools. E.g. when adding an ivar to the hierarchy, we have to recompile and this not announced, either. The recompile done when using e.g. metalinks is not announced, too.

The system would be unusable of we would force all tools to update for all these cases

**Comment by MarcusDenker** [comment]
> This is the number but also the order I guess? (Because we could change the index of the slots)

Ah, yes, you are right ! If you access ivars from now the newly changed superclass, then, yes, it might be that the offset is different due to the order of the ivars being different.

So we recompile all  fpor the the same reason why we recompile all if you add or remove and ivar: its the simplest. We could skip methods that do not access ivars, there is an issue for that: https://github.com/pharo-project/pharo/issues/14136

**Comment by MarcusDenker** [comment]
> methods reaching an inconsistent state. In our case, an index for speeding up senders and implementors in large images that stores compiled methods gets in an invalid state in case a method is recompiled in a class, and that recompilation is not announced...

That's why speeding up "senders of" is not that easy.

Your index can not have references to the CompiledMethod, but should have a "Class name/method name" pair an re-lookup the compiledMethod.

**Comment by chisandrei** [comment]
I definitely understand the speedup argument, and the need to not generate many announcements during bulk operations, like loading code. 

But also with a methods cache by class/selector there is price to pay in terms of memory and performance for doing the look-up. During normal code editing, the speed-up of not sending announcements is usually low.

We also use a mixed strategy now in many cases, when during code loading we disabled many announcements, use dedicated bulk loaders for code, and update the state of tools after the loading is finished.

The ideal case could be to have to have an extension to the class loader/system that would raise announcements anytime a compiled method is changed. That extension could be enabled by users explicitly when they would need those changes. Just for this particular change, we did not see an easy way to detect that subclasses should be updated.

Also about compiled methods, would the current direction be for tools to not really on them, and for the system to not announce just changes to compiled methods but have more high level events?

**Comment by chisandrei** [comment]
But maybe for this issue the better fix would indeed be https://github.com/pharo-project/pharo/issues/14136 :). 

Just in that case when a change to the instance variables of a superclass would cause the class layout of subclasses to be changed and methods to be recompiled, does that trigger announcements for subclasses that their layout changed?

**Comment by MarcusDenker** [comment]
it would not fix this issue, as the me methods accessing ivars have to be recompiled. and that still is not. a “all browsers, the method channged”, because it did not change.


the deeper problem is that CompildMethod is such a low level object. imagine the bytecode would be in a bytearray, then just that would  change.

**Comment by chisandrei** [comment]
In the case of modifying an instance variable in a superclass, we get `ClassModifiedClassDefinition` in all subclasses, which is very useful as we can then react and update caches holding compiled methods.

In this particular case, I do not see this as announcing individual changes when compiled methods are changed, but just having an announcement to indicate that because the superclass was changed, subclasses were rebuild. Maybe even a different announcement like `ClassModifiedClassRebuild`. The rebuild of a class is not a total transparent operation, as the identity of some objects changes afterwards. But maybe also since in the current builder changing a class definition results in the rebuild of subclasses, just hooking to the ClassModifiedClassDefinition of the subclass could be enough.

**Comment by chisandrei** [comment]
Maybe to make the conversation more concrete, on the above hierarchy we can just do a change of instance variables in class `MySubClass1`. In that case there are `ClassModificationApplied` announcements for each class in the hierarchy.

<img width="1088" alt="Screenshot 2023-10-16 at 22 43 34" src="https://github.com/pharo-project/pharo/assets/3534711/137dc082-b96d-4016-b762-884a71028140">

If we just change the superclass of `MySubClass1` only a single change is announced for `MySubClass1`

<img width="1088" alt="Screenshot 2023-10-16 at 22 44 26" src="https://github.com/pharo-project/pharo/assets/3534711/fde91196-cfef-46f1-916d-4cc9d21618c9">

Both changes modify the class definition and they both end up building new class versions for subclasses, but only one gets announced for subclasses.

**Comment by MarcusDenker** [comment]
We can announce a class change, yes. 

It's just a per method recompile notification that I think would be problematic.

**Comment by chisandrei** [comment]
>It's just a per method recompile notification that I think would be problematic.

Definitely. Re-reading the thread, I see that I referred too much to compiled methods. This issue is just about announcing the subclass changes.
